app/main/util/render_utils.py

```
@@ -72,53 +142,103 @@ def create_presigned_url_for_access_copy(file: File) -> str:
     return presigned_url
 
 
+def get_pdf_pages_from_s3(file_obj):
+    """
+    Fetch PDF file from S3, extract pages as images, and return page data.
+    Returns (page_data, error_response) where error_response is a Flask response or None.
+    """
+    try:
+        pdf_bytes = file_obj["Body"].read()
+        current_app.logger.info(f"PDF bytes length: {len(pdf_bytes)}")
+
+        page_data = extract_pdf_pages_as_images(pdf_bytes)
+        current_app.logger.info(f"Extracted {len(page_data)} pages from PDF")
+        return page_data, None
+    except Exception as e:
+        current_app.logger.error(
+            f"Error processing PDF for thumbnails: {e}", exc_info=True
+        )
+        return None, jsonify({"error": "Failed to process PDF"}), 500
```

This is entirely fine to do, it works, but we could just re raise the error rather than having the return signature being a tuple of (page_data, error_response).

Apart from generally being the more pythonic/standard approach, one major benefit is that even if we don't explicitly catch the raised error and do something with it in the flask route this gets called in, it will propagate and trigger flask

Whereas with this approach, the error could technically pass silently if we dont bother to handle it in the calling function.

See https://peps.python.org/pep-0020/#the-zen-of-python

This is unlike, say, Rust which uses something akin to a tuple https://doc.rust-lang.org/std/result/ but it enforces that the error is handled by its compiler.

Go similarly uses the result, err pattern, but for all functions. It doesnt enforce that you handle the errors but because thats the standard approach it makes sense.

Of course, python can do this, but it is not the standard convention and thus I suggest keeping to the standard for simplicity.

---

Done, change has been made

---

I still see it here. Let me know if you want to talk through it